### PR TITLE
[Feat] add impl for aggregate、 auto increment、 checkAndInsert. add demo

### PR DESCRIPTION
**Pull request by SmallSongZ**
### Task Description
[Feat] add impl for aggregate、 auto increment、 checkAndInsert. add demo

### Solution Description
add impl and demo

### Passed Regressions
Pass tests except for index-tests in 4.x

### Upgrade Compatibility
N.A

### Other Information
N.A

### CODE REVIEW

File: protocol/query.go [hunk 1]
```diff
@@ -221,6 +221,51 @@ func (q *ObTableQuery) SetAggregations(aggregations []*ObTableAggregationSingle)
...
+func (q *ObTableQuery) IsAggregations() bool {
+	if q.aggregations == nil {
+		return false
+	} else {
+		return true
+	}
+}
+
+// TransferQueryRange sets the query range into tableQuery.
+func (q *ObTableQuery) TransferQueryRange(rangePair []*table.RangePair) error {
```
Comment (WeiXinChan) [inline protocol/query.go:233]: to be common function

File: protocol/query_and_mutate.go [hunk 2]
```diff
@@ -47,13 +47,19 @@ func NewObTableQueryAndMutate() *ObTableQueryAndMutate {
...
 	obTableBatchOperation := NewObTableBatchOperation()
-	obTableOperation, err := NewObTableOperationWithParams(tableOperationType, nil, columns)
-	if err != nil {
-		return nil, errors.WithMessage(err, "create table operation")
+	if tableOperationType == ObTableOperationInsert {
+		obTableOperation, err := NewObTableOperationWithParams(tableOperationType, rowKey, columns)
+		if err != nil {
+			return nil, errors.WithMessage(err, "create table operation")
+		}
+		obTableBatchOperation.SetObTableOperations([]*ObTableOperation{obTableOperation})
```
Comment (WeiXinChan) [inline protocol/query_and_mutate.go:55]: obTableBatchOperation.SetObTableOperations([]*ObTableOperation{obTableOperation}) to be common line

File: client/aggregate_result.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+func newObAggregateResult(aggResultSet QueryResult) *obAggregateResult {
+	return &obAggregateResult{aggResultSet}
+}
+
+type obAggregateResult struct {
+	aggResult QueryResult
+}
+
+// Value returns the value of the specified column.
+func (r *obAggregateResult) Value(columnName string) interface{} {
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply client/aggregate_result.go:34]: fixed

File: client/aggregate_result.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+func newObAggregateResult(aggResultSet QueryResult) *obAggregateResult {
+	return &obAggregateResult{aggResultSet}
+}
+
+type obAggregateResult struct {
+	aggResult QueryResult
+}
+
+// Value returns the value of the specified column.
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply client/aggregate_result.go:33]: fixed

File: protocol/aggregation_single.go [hunk 5]
```diff
@@ -40,6 +40,17 @@ func NewObTableAggregationSingle() *ObTableAggregationSingle {
 	}
 }
 
+func NewObTableAggregationSingleWithParams(aggregationType ObTableAggregationType, aggregationColumn string) *ObTableAggregationSingle {
+	return &ObTableAggregationSingle{
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply protocol/aggregation_single.go:44]: fixed

File: protocol/query_and_mutate.go (hunk 2 above)
Comment (SmallSongZ) [reply protocol/query_and_mutate.go:55]: do not need to change

File: example/aggregate/aggregate.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	startRowKey := []*table.Column{table.NewColumn("c1", int64(0)), table.NewColumn("c2", table.Min)}
+	endRowKey := []*table.Column{table.NewColumn("c1", int64(100)), table.NewColumn("c2", table.Max)}
+	keyRanges := []*table.RangePair{table.NewRangePair(startRowKey, endRowKey)}
+
+	// set agg params
+	aggs := []*protocol.ObTableAggregationSingle{
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply example/aggregate/aggregate.go:50]: fixed, impl like batch executor

File: client/client.go [hunk 7]
```diff
@@ -87,6 +87,8 @@ type Client interface {
 	Get(ctx context.Context, tableName string, rowKey []*table.Column, getColumns []string, opts ...option.ObOperationOption) (SingleResult, error)
 	// Query records by rangePairs.
 	Query(ctx context.Context, tableName string, rangePairs []*table.RangePair, opts ...option.ObQueryOption) (QueryResultIterator, error)
+	// Aggregate records by rangePairs.
+	Aggregate(ctx context.Context, tableName string, rangePairs []*table.RangePair, opts ...option.ObQueryOption) (AggregateResult, error)
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply client/client.go:91]: fixed

File: client/obclient.go [hunk 8]
```diff
@@ -396,6 +411,24 @@ func (c *obClient) Query(ctx context.Context, tableName string, rangePairs []*ta
...
 }
 
+func (c *obClient) Aggregate(ctx context.Context, tableName string, rangePairs []*table.RangePair, opts ...option.ObQueryOption) (AggregateResult, error) {
+	queryOpts := c.getObQueryOptions(opts...)
+	queryExecutor := newObQueryExecutorWithParams(tableName, c)
+	queryExecutor.addKeyRanges(rangePairs)
+	queryExecutor.setQueryOptions(queryOpts)
+	resSet, err := queryExecutor.init(ctx)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "execute aggregate, tableName:%s, keyRanges:%s",
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply client/obclient.go:421]: fixed

File: client/aggregate_result.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+func (r *obAggregateResult) Value(columnName string) interface{} {
+	if r.aggResult == nil {
+		return nil
+	}
+	return r.aggResult.Value(columnName)
+}
+
+// Values returns all values in the query result.
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply client/aggregate_result.go:41]: fixed

File: client/aggregate_result.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+ * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
+ * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+ * See the Mulan PSL v2 for more details.
+ * #L%
+ */
+
+package client
+
+type AggregateResult interface {
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply client/aggregate_result.go:20]: fixed

File: protocol/query.go (hunk 1 above)
Comment (SmallSongZ) [reply protocol/query.go:233]: add this function in client/table.go

File: example/filter_operation/insert.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	// set key range
+	startRowKey := []*table.Column{table.NewColumn("c1", "1")}
+	endRowKey := []*table.Column{table.NewColumn("c1", "10")}
+	keyRanges := []*table.RangePair{table.NewRangePair(startRowKey, endRowKey)}
+
+	rowKey = []*table.Column{table.NewColumn("c1", "4")}
+	mutationColumns := []*table.Column{table.NewColumn("c2", int64(3))}
+
+	// satisfy the filter, insert (4, 3)
+	affectRows, err = cli.Insert(
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply example/filter_operation/insert.go:72]: done

File: example/auto_increment/auto_inc_rowkey.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	cli, err := client.NewClient(configUrl, fullUserName, passWord, sysUserName, sysPassWord, cfg)
+	if err != nil {
+		panic(err)
+	}
+
+	// insert. set 0, use auto increment value
+	// insert (1, 1, hello, 0)
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second) // 10s
+	rowKey := []*table.Column{table.NewColumn("c1", int64(0)), table.NewColumn("c2", int64(1))}
+	affectRows, err := cli.Insert(
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_rowkey.go:50]: done

File: example/auto_increment/auto_inc_not_rowkey.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	cli, err := client.NewClient(configUrl, fullUserName, passWord, sysUserName, sysPassWord, cfg)
+	if err != nil {
+		panic(err)
+	}
+
+	// insert. set 0, use auto increment value
+	// insert (1, 1, hello, 0)
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second) // 10s
+	rowKey := []*table.Column{table.NewColumn("c1", int64(1))}
+	affectRows, err := cli.Insert(
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_not_rowkey.go:50]: done

File: client/aggregate_executor.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+package client
+
+import (
+	"context"
+)
+
+type AggExecutor interface {
+	// Min add a min operation to the agg executor.
+	Min(aggColumn string) AggExecutor
```
Comment (WeiXinChan) [inline client/aggregate_executor.go:26]: rename aggColumn -> columnName

File: client/obaggregate.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+	return
+}
+
+// Execute an agg operation.
+// AggregateResult contains the results of all operations.
+func (q *obAggExecutor) Execute(ctx context.Context) (AggregateResult, error) {
+	if len(q.aggOperations) == 0 {
+		return nil, errors.New("empty aggregation operations")
+	}
+	q.setParams()
```
Comment (WeiXinChan) [inline client/obaggregate.go:88]: no need to use a method
q.queryExecutor.tableQuery.SetSelectColumns(q.aggOperationNames)
q.queryExecutor.tableQuery.SetAggregations(q.aggOperations)

File: client/table.go [hunk 16]
```diff
@@ -108,3 +109,39 @@ func (t *ObTable) String() string {
 		"isClosed:" + strconv.FormatBool(t.isClosed) +
 		"}"
 }
+
+// transferQueryRange sets the query range into tableQuery.
+func transferQueryRange(rangePair []*table.RangePair) ([]*protocol.ObNewRange, error) {
```
Comment (WeiXinChan) [inline client/table.go:114]: move to range_pair.go and add unit test

File: example/aggregate/aggregate.go [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	startRowKey := []*table.Column{table.NewColumn("c1", int64(0)), table.NewColumn("c2", table.Min)}
+	endRowKey := []*table.Column{table.NewColumn("c1", int64(100)), table.NewColumn("c2", table.Max)}
+	keyRanges := []*table.RangePair{table.NewRangePair(startRowKey, endRowKey)}
+
+	// create agg executor
+	aggExecutor := cli.NewAggExecutor(tableName, keyRanges).Sum("c3")
```
Comment (WeiXinChan) [inline example/aggregate/aggregate.go:48]: aggExecutor := cli.NewAggExecutor(tableName, keyRanges).Sum("c3").Min("c1").Max("c2")

File: example/aggregate/aggregate.go [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+	// get agg result
+	res, err := aggExecutor.Execute(context.TODO())
```
Comment (WeiXinChan) [inline example/aggregate/aggregate.go:51]: check error

Comment (WeiXinChan) [reply example/aggregate/aggregate.go:51]: you have to show a good example to user

File: example/auto_increment/auto_inc_not_rowkey.go [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+		rowKey,
+		nil,
+	)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(affectRows)
+
+	// get and check
+	startRowKey := []*table.Column{table.NewColumn("c1", int64(1))}
```
Comment (WeiXinChan) [inline example/auto_increment/auto_inc_not_rowkey.go:63]: use get

File: example/auto_increment/auto_inc_not_rowkey.go [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+		panic(err)
+	}
+	fmt.Println(affectRows)
+
+	// get and check
+	startRowKey = []*table.Column{table.NewColumn("c1", int64(2))}
+	endRowKey = []*table.Column{table.NewColumn("c1", int64(2))}
+	keyRanges = []*table.RangePair{table.NewRangePair(startRowKey, endRowKey)}
+
+	resSet, err = cli.Query(
```
Comment (WeiXinChan) [inline example/auto_increment/auto_inc_not_rowkey.go:103]: use get

File: example/auto_increment/auto_inc_not_rowkey.go [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+		tableName,
+		keyRanges,
+		option.WithSelectColumns([]string{"c1", "c2", "c3"}),
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	res, err = resSet.Next()
```
Comment (WeiXinChan) [inline example/auto_increment/auto_inc_not_rowkey.go:114]: check error anytime

File: example/auto_increment/auto_inc_rowkey.go [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+		tableName,
+		keyRanges,
+		option.WithSelectColumns([]string{"c1", "c2", "c3"}),
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	res, err = resSet.Next()
```
Comment (WeiXinChan) [inline example/auto_increment/auto_inc_rowkey.go:113]: check error anytime

File: example/filter_operation/insert.go [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+		panic(err)
+	}
+	fmt.Println(affectRows)
+
+	// get and check
+	startRowKey = []*table.Column{table.NewColumn("c1", int64(4))}
+	endRowKey = []*table.Column{table.NewColumn("c1", int64(4))}
+	keyRanges = []*table.RangePair{table.NewRangePair(startRowKey, endRowKey)}
+
+	resSet, err := cli.Query(
```
Comment (WeiXinChan) [inline example/filter_operation/insert.go:90]: use get

File: example/filter_operation/insert.go [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+		tableName,
+		keyRanges,
+		option.WithSelectColumns([]string{"c1", "c2"}),
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	res, err := resSet.Next()
```
Comment (WeiXinChan) [inline example/filter_operation/insert.go:101]: check error

File: client/aggregate_executor.go (hunk 14 above)
Comment (SmallSongZ) [reply client/aggregate_executor.go:26]: fixed

File: client/obaggregate.go (hunk 15 above)
Comment (SmallSongZ) [reply client/obaggregate.go:88]: fixed

File: example/auto_increment/auto_inc_rowkey.go (hunk 22 above)
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_rowkey.go:113]: fixed

File: example/filter_operation/insert.go (hunk 23 above)
Comment (SmallSongZ) [reply example/filter_operation/insert.go:90]: fixed

File: example/filter_operation/insert.go (hunk 24 above)
Comment (SmallSongZ) [reply example/filter_operation/insert.go:101]: fixed

File: example/auto_increment/auto_inc_rowkey.go [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+		panic(err)
+	}
+	fmt.Println(affectRows)
+
+	// get and check
+	startRowKey = []*table.Column{table.NewColumn("c1", int64(50))}
+	endRowKey = []*table.Column{table.NewColumn("c1", int64(50))}
+	keyRanges = []*table.RangePair{table.NewRangePair(startRowKey, endRowKey)}
+
+	resSet, err = cli.Query(
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_rowkey.go:102]: fixed

File: example/auto_increment/auto_inc_rowkey.go [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+		tableName,
+		keyRanges,
+		option.WithSelectColumns([]string{"c1", "c2", "c3"}),
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	res, err := resSet.Next()
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_rowkey.go:78]: fixed

File: example/aggregate/aggregate.go (hunk 18 above)
Comment (SmallSongZ) [reply example/aggregate/aggregate.go:51]: fixed

File: client/table.go (hunk 16 above)
Comment (SmallSongZ) [reply client/table.go:114]: add util.go in client

File: example/aggregate/aggregate.go (hunk 17 above)
Comment (SmallSongZ) [reply example/aggregate/aggregate.go:48]: fixed

File: example/auto_increment/auto_inc_not_rowkey.go (hunk 19 above)
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_not_rowkey.go:63]: fixed

File: example/auto_increment/auto_inc_not_rowkey.go (hunk 20 above)
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_not_rowkey.go:103]: fixed

File: example/auto_increment/auto_inc_not_rowkey.go (hunk 21 above)
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_not_rowkey.go:114]: fixed

File: example/auto_increment/auto_inc_rowkey.go [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+		panic(err)
+	}
+	fmt.Println(affectRows)
+
+	// get and check
+	startRowKey := []*table.Column{table.NewColumn("c1", int64(1))}
+	endRowKey := []*table.Column{table.NewColumn("c1", int64(1))}
+	keyRanges := []*table.RangePair{table.NewRangePair(startRowKey, endRowKey)}
+
+	resSet, err := cli.Query(
```
[the remark this answers is not in the document]
Comment (SmallSongZ) [reply example/auto_increment/auto_inc_rowkey.go:67]: fixed

**Comment by WeiXinChan** [comment]
LGTM

**Pull request merged**